### PR TITLE
[fix][broker] Added check for invisible characters for subscription name

### PR DESCRIPTION
**Pull request by lordcheng10**
### Motivation
When the subscription name contains some invisible characters, such as \u0000, the subscription cannot be deleted：
image

image

So when creating a subscription name, we should check the subscription name: create subscription group fails when subscription name contains invisible characters

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [x] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### Matching PR in forked repository

PR in forked repository: https://github.com/lordcheng10/pulsar/pull/3

### CODE REVIEW

**Comment by lordcheng10** [comment]
@codelipenghui @Technoboy- PTAL,thanks!

**Comment by equanz** [comment]
I think we can't resolve the issue("the subscription cannot be deleted") entirely by this approach. Because we can't unsubscribe existing subscriptions, also this approach can only affect REST API.

Moreover, I think the current restriction is too strict for me. For example, I want to use dots for subscription names. Could you modify this restriction to be able to enable/disable it?

```
% ./bin/pulsar-admin topics create-subscription -s api.sub persistent://public/default/topic
2022-08-24T18:00:31,752+0900 [AsyncHttpClient-7-1] WARN  org.apache.pulsar.client.admin.internal.BaseResource - [http://localhost:8080/admin/v2/persistent/public/default/topic/subscription/api.sub?replicated=false] Failed to perform http put request: javax.ws.rs.BadRequestException: HTTP 400 subscriptionName contains special characters
subscriptionName contains special characters

Reason: subscriptionName contains special characters
```

**Comment by lordcheng10** [comment]
> I think we can't resolve the issue("the subscription cannot be deleted") entirely by this approach. Because we can't unsubscribe existing subscriptions, also this approach can only affect REST API.
> 
> Moreover, I think the current restriction is too strict for me. For example, I want to use dots for subscription names. Could you modify this restriction to be able to enable/disable it?
> 
> ```
> % ./bin/pulsar-admin topics create-subscription -s api.sub persistent://public/default/topic
> 2022-08-24T18:00:31,752+0900 [AsyncHttpClient-7-1] WARN  org.apache.pulsar.client.admin.internal.BaseResource - [http://localhost:8080/admin/v2/persistent/public/default/topic/subscription/api.sub?replicated=false] Failed to perform http put request: javax.ws.rs.BadRequestException: HTTP 400 subscriptionName contains special characters
> subscriptionName contains special characters
> 
> Reason: subscriptionName contains special characters
> ```

IMO,
1. First, the subscription name should not contain special characters, and the specific naming rules can be discussed;
2. For the existing illegal subscription name, it can be deleted by dynamically opening the subscription name expiration cleanup, but it cannot be deleted by command, because invisible characters (eg: /u0000) cannot be input by command；

@equanz @codelipenghui @HQebupt @Technoboy- PTAL,thanks!

**Comment by lordcheng10** [comment]
@Jason918 PTAL,thanks!

**Comment by massakam** [comment]
This is a breaking change and may prevent existing users from upgrading Pulsar. For example, using host names as subscriptions is a very common use case for our company. Host names commonly contain multiple `.` characters, but this PR makes it impossible to create subscriptions containing `.`. Therefore, this PR is unacceptable for us.

At least, a flag should be added to the broker's configuration to enable/disable this check. Moreover, the current check is too strict. For example, even if a subscription contains `.`, we can remove it without any problem. So there is no need to prohibit the creation of such subscriptions.

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.utils;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringUtil {
+    private static String regEx = "[ _`~!@#$%^&*()+=|{}':;',\\[\\].<>/?~！@#￥%……&*（）——+|{}【】‘；：”“’。，、？]|\n|\r|\t";
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:25]: Is there some lib that can check if a char is printable or visible?
IMO, '.' should be here.

BTW, does topic name have the same issue?

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:25]: > BTW, does topic name have the same issue?

Makes sense, I'll check

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:25]: > Is there some lib that can check if a char is printable or visible?
IMO, '.' should be here.

OK, I agree to only check for invisible characters

**Comment by lordcheng10** [comment]
> This is a breaking change and may prevent existing users from upgrading Pulsar. For example, using host names as subscriptions is a very common use case for our company. Host names commonly contain multiple `.` characters, but this PR makes it impossible to create subscriptions containing `.`. Therefore, this PR is unacceptable for us.
> 
> At least, a flag should be added to the broker's configuration to enable/disable this check. Moreover, the current check is too strict. For example, even if a subscription contains `.`, we can remove it without any problem. So there is no need to prohibit the creation of such subscriptions.

OK, I agree to only check for invisible characters

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java (hunk 1 above)
Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: I didn't find some lib that can check if a char is printable or visible.
But I found that the decimal number of invisible characters is 0~31, so I added a method org.apache.pulsar.utils.StringUtil#invisibleCharacters to check if the subscription name contains invisible characters.

image

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: @Jason918 PTAL,thanks!

I didn't find some lib that can check if a char is printable or visible.
The ASCII values of invisible characters are 0~32 and 127, so I added a method org.apache.pulsar.utils.StringUtil#invisibleCharacters to check if the subscription name contains invisible characters.
image

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: > Is there some lib that can check if a char is printable or visible? IMO, '.' should be here.
> 
> BTW, does topic name have the same issue?

@Jason918 PTAL,thanks!

I didn't find some lib that can check if a char is printable or visible.
The ASCII values of invisible characters are 0~32 and 127, so I added a method org.apache.pulsar.utils.StringUtil#invisibleCharacters to check if the subscription name contains invisible characters.
image

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: > BTW, does topic name have the same issue?

Yes，I added a check for invisible characters in topic name！

PTAL，thanks! @Jason918

Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: @lordcheng10 The problem is that Java uses Unicode by default. There are a lot of helper method in `java.lang.Character`, Please take a look at the methods like `isWhitespace(char)`  `getType(char)`. I am not certain if we can cover all the cases easily. It's worth a try.

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: isWhitespace does not cover all invisible characters

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: @Jason918 
fixed! PTAL,thanks!

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java]: Invisible characters are generally of these types:
                Character.CONTROL:
                Character.FORMAT:
                Character.PRIVATE_USE:
                Character.SURROGATE:
                Character.UNASSIGNED:
                Character.OTHER_SYMBOL:
                Character.SPACE_SEPARATOR:

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.utils;
+
+import org.apache.pulsar.common.naming.TopicName;
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:21]: Seems not relevant？

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+public class StringUtil {
+
+    /**
+     * Check for invisible characters.
+     * */
+    public static boolean invisibleCharacters(String str) {
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:28]: `containsInvisibleCharacters` is more expressive?

File: pulsar-broker/src/test/java/org/apache/pulsar/utils/StringUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.assertFalse;
+import org.testng.annotations.Test;
+
+
+@Test(groups = "utils")
+public class StringUtilTest {
+
+    @Test
+    public void testInvisibleCharacters() {
```
Comment (Jason918) [inline pulsar-broker/src/test/java/org/apache/pulsar/utils/StringUtilTest.java:30]: Add other typical invisible cases like ' ', '\n' etc?
Maybe add case contains Chinese? 
And I wonder if this works for Chinese space '\u3000'?

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java (hunk 2 above)
Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:21]: fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java (hunk 3 above)
Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/StringUtil.java:28]: fixed

File: pulsar-broker/src/test/java/org/apache/pulsar/utils/StringUtilTest.java (hunk 4 above)
Comment (lordcheng10) [reply pulsar-broker/src/test/java/org/apache/pulsar/utils/StringUtilTest.java]: fixed

**Review by Jason918 — APPROVED** [review]
LGTM

**Comment by lordcheng10** [comment]
@eolivelli @codelipenghui @hangc0276 @Technoboy- PTAL,thanks!

**Review by poorbarcode — COMMENTED** [review]
Can you add a test for the admin API

**Comment by lordcheng10** [comment]
> Can you add a test for the admin API ?

fixed. PTAL,thanks! @poorbarcode

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/CreateSubscriptionTest.java [hunk 5]
```diff
@@ -135,6 +135,46 @@ public void createSubscriptionOnPartitionedTopic() throws Exception {
...
+        admin.topics().createSubscription(topic, "sub-1", MessageId.latest);
+        admin.topics().createSubscription(topic, "sub~`!@#$￥%^……&*(){「【[]}】』\\|、:;\"'<《,>。》.?/-1",
+                MessageId.latest);
+
+        // Create should fail if the subscription name contain invisible characters
+        try {
+            admin.topics().createSubscription(topic, "sub\u0000-2", MessageId.latest);
+            fail("Should have failed");
+        } catch (Exception e) {
+            // Expected
```
Comment (poorbarcode) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/CreateSubscriptionTest.java:150]: could we check the response status and error message?

Comment (lordcheng10) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/CreateSubscriptionTest.java]: Fixed! @poorbarcode

**Review by codelipenghui — APPROVED** [review]

**Comment by lordcheng10** [comment]
@Technoboy- PTAL,thanks!

**Review by michaeljmarshall — COMMENTED** [review]
I think the change makes sense. Is there a reason we don't do this validation for subscriptions created via auto subscription creation in the Pulsar Protocol?

**Comment by codelipenghui** [comment]
> I think the change makes sense. Is there a reason we don't do this validation for subscriptions created via auto subscription creation in the Pulsar Protocol?

We should add the validation. It's better to keep consistent with the REST API and Pulsar Protocol.

**Comment by lordcheng10** [comment]
> > I think the change makes sense. Is there a reason we don't do this validation for subscriptions created via auto subscription creation in the Pulsar Protocol?
> 
> We should add the validation. It's better to keep consistent with the REST API and Pulsar Protocol.

OK, I will add this validation for subscriptions created via auto subscription creation in the Pulsar Protocol

**Comment by lordcheng10** [comment]
> I think the change makes sense. Is there a reason we don't do this validation for subscriptions created via auto subscription creation in the Pulsar Protocol?

Fixed.
Add this validation when automatically creating a subscription. @michaeljmarshall

**Review by lhotari — CHANGES_REQUESTED** [review]
I think that `containsInvisibleCharacters` is too limited. There might be other character classes that aren't suitable, so calling them "invisible" would be wrong.

Instead of checking for invalid characters, it would be better to have the check the validity of the given subscription name explicitly. 

Pulsar already contains org.apache.pulsar.common.naming.NamedEntity#checkName method and a similar solution could be followed in this case. I guess switching to use NamedEntity.checkName could be a breaking change for existing applications. Adding a specific solution for Subscription name validation would be justified.

It should also be discussed whether there's a need for a feature flag for enforcing a stricter validation of the subscription name with NamedEntity.checkName so that there would be consistency across other named entitiies in Pulsar.

**Comment by lordcheng10** [comment]
CI PR: https://github.com/lordcheng10/pulsar/pull/3

**Comment by lordcheng10** [comment]
> Is the root of the problem here that the character sequences in question are not being properly encoded when forming the admin URLs? From the example given, I am seeing what looks to be a URL encoding of a Java[Script] escaped form of the original name, rather than a direct encoding of the original name. There are also some pretty odd character codes in the formed URL — the backslash of `\u....` seems to have been encoded as `%255C`:
> 
> * [%255C — ╜](https://www.htmlsymbols.xyz/unicode/U+255c)
> * [%25BD — ▽](https://www.htmlsymbols.xyz/unicode/U+25bd)
> * [%25BF — ▿](https://www.htmlsymbols.xyz/unicode/U+25bf)
> * [%25ED — ◯](https://www.htmlsymbols.xyz/unicode/U+25ed)
> 
> Practically speaking, it should be possible to correctly transport any characters (including invisible ones) via a correctly encoded URL. I suspect there is a character encoding issue somewhere. If this is the case, I'd recommend:
> 
> 1. Preventing **new** names from being created with invisible/indistinguishable characters
> 2. Fix name encoding / URL handling so that any existing names with such characters can still be managed

1. There is no problem with encoding;
2. The problem is that the subscription name contains invisible characters, and invisible characters cannot be entered on the command line, resulting in the command reporting an error: Subscription not found

**Comment by lordcheng10** [comment]
> Instead of checking for invalid characters, it would be better to have the check the validity of the given subscription name explicitly.

I'll start a discussion: Naming rules for subscriptions. @lhotari

**Comment by lordcheng10** [comment]
> > Instead of checking for invalid characters, it would be better to have the check the validity of the given subscription name explicitly.
> 
> I'll start a discussion about naming conventions for subscriptions. @lhotari

discuss thread: https://lists.apache.org/thread/mbm8p3wh7k0bc3nxg3qq6tdbqfotckr5

**Comment by lordcheng10** [comment]
CI all passed: https://github.com/lordcheng10/pulsar/pull/3

**Comment by lordcheng10** [comment]
> I think that `containsInvisibleCharacters` is too limited. There might be other character classes that aren't suitable, so calling them "invisible" would be wrong.
> 
> Instead of checking for invalid characters, it would be better to have the check the validity of the given subscription name explicitly.
> 
> Pulsar already contains org.apache.pulsar.common.naming.NamedEntity#checkName method and a similar solution could be followed in this case. I guess switching to use NamedEntity.checkName could be a breaking change for existing applications. Adding a specific solution for Subscription name validation would be justified.
> 
> It should also be discussed whether there's a need for a feature flag for enforcing a stricter validation of the subscription name with NamedEntity.checkName so that there would be consistency across other named entitiies in Pulsar.

PTAL,thanks! @lhotari discuss thread: https://lists.apache.org/thread/mbm8p3wh7k0bc3nxg3qq6tdbqfotckr5

**Review by HQebupt — APPROVED** [review]

**Comment by tisonkun** [comment]
> and invisible characters cannot be entered on the command line

This should not be a reason. Users can input invisible characters in multiple ways. If the REST API can encode the name properly, then I'll regard that here's no issue to be "fixed".

**Comment by tisonkun** [comment]
Closed as no consensus.

I see @mattisonchao started a discussion about topic names on the mailing list: https://lists.apache.org/thread/5s4kop7qmxgpvsnh65s42r9mv7qc1pxt

@lordcheng10 if you'd like to continue this work, you can follow this pattern to discuss and get more attention first.

**Pull request closed without merging**